source/common/router/router_ratelimit.cc

```diff
@@ -109,28 +109,31 @@ void RateLimitPolicyEntryImpl::populateDescriptors(
 
 RateLimitPolicyImpl::RateLimitPolicyImpl(const Json::Object& config) {
   if (config.hasObject("rate_limits")) {
-    std::vector<std::unique_ptr<RateLimitPolicyEntry>> rate_limit_policy;
-    std::vector<std::reference_wrapper<const RateLimitPolicyEntry>> rate_limit_policy_reference;
+    // The supported number of stages is 0 through RateLimitPolicyImpl::MAX_STAGE_NUMBER
+    // inclusively.
+    for (size_t i = 0; i <= RateLimitPolicyImpl::MAX_STAGE_NUMBER; i++) {
+      std::vector<std::unique_ptr<RateLimitPolicyEntry>> rate_limit_policy;
+      std::vector<std::reference_wrapper<const RateLimitPolicyEntry>> rate_limit_policy_reference;
+      rate_limit_entries_reference_.emplace_back(rate_limit_policy_reference);
+      rate_limit_entries_.emplace_back(std::move(rate_limit_policy));
```

Comment:
Just resize() both vectors to the correct size, that should default construct empty vectors for each outer vector entry.